src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs

```diff
@@ -262,55 +267,91 @@ void StoreElement(PointerValue pointer, IValue value, string? fromLocalId, bool
-                var forceCopy = sharedState.Context.CreateConstant(false);
-                var copy = sharedState.CurrentBuilder.Call(createShallowCopy, originalArray.OpaquePointer, forceCopy);
-                var array = sharedState.Values.FromArray(copy, originalArray.QSharpElementType);
-                sharedState.ScopeMgr.RegisterValue(array);
+                ArrayValue GetArrayCopy(bool needsToBeCopied)
+                {
+                    // Since we keep track of alias counts for arrays we always ask the runtime to create a shallow copy
+                    // if needed. The runtime function ArrayCopy creates a new value with reference count 1 if the current
+                    // alias count is larger than 0, and otherwise merely increases the reference count of the array by 1.
+                    var createShallowCopy = sharedState.GetOrCreateRuntimeFunction(RuntimeLibrary.ArrayCopy);
+                    var forceCopy = sharedState.Context.CreateConstant(needsToBeCopied);
+                    var copy = sharedState.CurrentBuilder.Call(createShallowCopy, originalArray.OpaquePointer, forceCopy);
+                    return sharedState.Values.FromArray(copy, originalArray.QSharpElementType);
+                }
 
-                void UpdateElement(Func<Value, IValue> getNewItemForIndex, Value index)
+                void UpdateElement(Func<IValue> getNewItemForIndex, PointerValue itemToUpdate)
                 {
-                    var elementPtr = array.GetArrayElementPointer(index);
+                    var newElement = getNewItemForIndex();
+
                     if (unreferenceOriginal)
                     {
-                        sharedState.ScopeMgr.DecreaseReferenceCount(elementPtr);
+                        sharedState.ScopeMgr.IncreaseReferenceCount(newElement);
+                        sharedState.ScopeMgr.DecreaseReferenceCount(itemToUpdate);
                     }
 
-                    var newElement = getNewItemForIndex(index);
-                    StoreElement(elementPtr, newElement, fromLocalId);
+                    StoreElement(itemToUpdate, newElement, fromLocalId);
                 }
 
+                ArrayValue array;
                 if (accEx.ResolvedType.Resolution.IsInt)
                 {
+                    array = GetArrayCopy(false);
+
                     // do not increase the ref count here - we will increase the ref count of all new items at the end
                     IValue newItemValue = sharedState.EvaluateSubexpression(updated);
                     var index = sharedState.EvaluateSubexpression(accEx);
-                    UpdateElement(_ => newItemValue, index.Value);
+                    UpdateElement(() => newItemValue, array.GetArrayElementPointer(index.Value));
                 }
                 else if (accEx.ResolvedType.Resolution.IsRange)
                 {
+                    // fixme: if fromLocalId equals the lhs then we need to make a deep copy of updated ...
```

Comment:
Should this fixme be tracked as an issue?

Reply:
This comment is just a remainder of my notes, and I have revised the comment here to be a description of what is done.